### PR TITLE
Implement Client.Download

### PR DESCRIPTION
**Pull request by lmars**
An early stage PR to get feedback on the approach and for suggestions on appropriate tests to add.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -379,16 +379,31 @@ func (c *Client) Files() data.Files {
...
+	if !ok {
+		return ErrNotFound{name}
+	}
+	r, err := c.remote.Get("targets/"+name, localMeta.Length)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	s := io.TeeReader(io.LimitReader(r, localMeta.Length), dest)
+	actual, err := util.GenerateFileMeta(s)
```
Comment (titanous) [inline client/client.go]: Might want to add a comment that the copy/hash happens in here.

**Comment by titanous** [comment]
We need a `Targets` accessor method that provides the complete list of available targets.

**Comment by titanous** [comment]
Things we should test:

- Target that is too long doesn't get read fully
- Target that is too short errors
- Target with incorrect hash errors

**Comment by lmars** [comment]
@titanous comments addressed

File: client/client.go [hunk 2]
```diff
@@ -379,16 +379,68 @@ func (c *Client) Files() data.Files {
...
-		TUF downloads and verifies the file and then makes the file available to the
-		software update system.
-	*/
 	return nil
 }
+
+// Targets returns the complete list of available targets.
+func (c *Client) Targets() (data.Files, error) {
+	// populate c.localTargets from local storage
+	if err := c.getLocalMeta(); err != nil {
```
Comment (titanous) [inline client/client.go]: Check if `localTargets` is set before doing this?

File: client/client_test.go [hunk 3]
```diff
@@ -182,3 +209,81 @@ func (s *ClientSuite) TestNewTargets(c *C) {
...
+}
+
+func (s *ClientSuite) TestDownloadTargetTooLong(c *C) {
+	client := s.updatedClient(c)
+	remoteFile := s.remote["targets/foo.txt"]
+	remoteFile.buf = bytes.NewReader([]byte("foo-ooo"))
+	var dest testDestination
+	c.Assert(client.Download("foo.txt", &dest), IsNil)
+	c.Assert(remoteFile.bytesRead, Equals, 3)
+	c.Assert(dest.deleted, Equals, false)
```
Comment (titanous) [inline client/client_test.go]: Should this be true?

Comment (lmars) [inline client/client_test.go]: So I guess this is a bit of an edge case. Usually the remote file would have a size corresponding to size of the data, which would result in an `ErrWrongSize` error and the stream being closed immediately (see [this added test](https://github.com/flynn/go-tuf/blob/client-download/client/client_test.go#L248-L256)). Here though the remote size is correct, so we just read the correct number of bytes, get the correct data and all is good, does it then matter that the remote data had more bytes?

**Comment by lmars** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
LGTM

File: client/client_test.go (hunk 3 above)
Comment (heavenlyhash) [inline client/client_test.go]: It's hard to think of a scenario where something that odd happens and it's still correct, so I'd agree with rejecting that state.

Comment (titanous) [inline client/client_test.go]: If the length is the same and the content hashes correctly, does it matter?

Comment (lmars) [inline client/client_test.go]: FYI this is the same behaviour as [the Python implementation](https://github.com/theupdateframework/tuf/blob/develop/tests/test_download.py#L108-L113)
